tutorials/how_to/span_representations.md

```diff
@@ -0,0 +1,180 @@
+                                      [3, 4]]]))
+
+# We concatenated the representations for the start and end of
+# the span, so the embedded span size is 2 * embedding_size.
+# shape (batch_size, num_spans, 2 * embedding_size).
+span_representations = extractor(sequence_tensor, indices)
+assert list(span_representations.size()) == [2, 2, 14]
+```
+
+There are other types of Span Extractors - for instance, the `SelfAttentiveSpanExtractor`,
+which computes span representations by generating an unnormalized attention score for each
+word in the sentence. Spans representations are then computed with respect to these
+scores by normalising the attention scores for words inside the span.
+
+
+## Scoring and Pruning Spans
+
+Span based representations have been effective for modelling/approximating structured
+prediction problems - however, many models which leverage this type of representation
+also involve some kind of _span enumeration_ (i.e considering all possible spans in a 
+sentence/document). For a given sentence of length n, there are n^2 spans. In itself, 
+this is not too problematic, but for instance, the co-reference model in AllenNLP 
+compares _pairs_ of spans - meaning that naively we consider n^4 spans, with potential
+document lengths of upwards of 3000 tokens.
+
+In order to solve this problem, we need to be able to _prune_ spans as we go inside our
+model. There are several ways to do this:
+
+### Heuristically prune spans in your DatasetReader.
+
+We have added a utility method for enumerating all spans in a sentence, but excluding
+those which fulfil some condition based on the input text or any Spacy `Token` attribute.
+For instance, for co-reference, all spans which are mentions (spans which are 
+co-referent with _something_) never start or end with punctuation, because of the way
+the Onotonotes 5.0 dataset was created. This means that we can exclude any span which
+starts or ends with punctuation using a very simple python function:
+
+```
+from typing import List
+from allennlp.data.dataset_readers.dataset_utils import span_utils
+from allennlp.data.tokenizers.word_tokenizer import SpacyWordSplitter
+from allennlp.data.tokenizers.token import Token
+
+tokenizer = SpacyWordSplitter(pos_tags=True)
+sentence = tokenizer.split_words("This is a sentence.")
+
+def no_prefixed_punctuation(tokens: List[Token]) -> bool:
+    # Only include spans which don't start or end with punctuation.
+    return tokens[0].pos_ != "PUNCT" and tokens[-1].pos_ != "PUNCT"
+
+spans = span_utils.enumerate_spans(sentence,
+                                    max_span_width=3,
+                                    min_span_width=2,
+                                    filter_function=no_prefixed_punctuation)
+
+# 'spans' won't include (2, 4) or (3, 4) as these have
+# punctuation as their last element. Note that these spans
+# have inclusive start and end indices!
+assert spans == [(0, 1), (0, 2), (1, 2), (1, 3), (2, 3)]
```

Comment:
One argument for putting this above when you talk about `DatasetReaders`: you could actually give a full example where you use this function and create a `ListField[SpanField]`.